integration/confidential/lib.sh

```diff
@@ -0,0 +1,177 @@
+#
+# Environment variables:
+#	RUNTIME_CONFIG_PATH - path to kata's configuration.toml. If it is not
+#			      export then it will figure out the path via
+#			      `kata-runtime env` and export its value.
+#
+add_kernel_params() {
+	local params="$@"
+	load_RUNTIME_CONFIG_PATH
+
+	sudo sed -i -e 's#^\(kernel_params\) = "\(.*\)"#\1 = "\2 '"$params"'"#g' \
+		"$RUNTIME_CONFIG_PATH"
+}
+
+# Get the 'kernel_params' property on kata's configuration.toml
+#
+# Environment variables:
+#	RUNTIME_CONFIG_PATH - path to kata's configuration.toml. If it is not
+#			      export then it will figure out the path via
+#			      `kata-runtime env` and export its value.
+#
+get_kernel_params() {
+	load_RUNTIME_CONFIG_PATH
+
+        local kernel_params=$(sed -n -e 's#^kernel_params = "\(.*\)"#\1#gp' \
+                "$RUNTIME_CONFIG_PATH")
+	echo "$kernel_params"
+}
+
+# Clear the 'kernel_params' property on kata's configuration.toml
+#
+# Environment variables:
+#	RUNTIME_CONFIG_PATH - path to kata's configuration.toml. If it is not
+#			      export then it will figure out the path via
+#			      `kata-runtime env` and export its value.
+#
+clear_kernel_params() {
+	load_RUNTIME_CONFIG_PATH
+
+	sudo sed -i -e 's#^\(kernel_params\) = "\(.*\)"#\1 = ""#g' \
+		"$RUNTIME_CONFIG_PATH"
+}
+
+# Enable the agent console so that one can open a shell with the guest VM.
+#
+# Environment variables:
+#	RUNTIME_CONFIG_PATH - path to kata's configuration.toml. If it is not
+#			      export then it will figure out the path via
+#			      `kata-runtime env` and export its value.
+#
+enable_agent_console() {
+	load_RUNTIME_CONFIG_PATH
+
+	sudo sed -i -e 's/^# *\(debug_console_enabled\).*=.*$/\1 = true/g' \
+		"$RUNTIME_CONFIG_PATH"
+}
+
+enable_full_debug() {
+	# Load the RUNTIME_CONFIG_PATH variable.
+	load_RUNTIME_CONFIG_PATH
+
+	# Toggle all the debug flags on in kata's configuration.toml to enable full logging.
+	sudo sed -i -e 's/^# *\(enable_debug\).*=.*$/\1 = true/g' "$RUNTIME_CONFIG_PATH"
+
+	# Also pass the initcall debug flags via Kernel parameters.
+	add_kernel_params "agent.log=debug" "initcall_debug"
+}
+
+disable_full_debug() {
+	# Load the RUNTIME_CONFIG_PATH variable.
+	load_RUNTIME_CONFIG_PATH
+
+	# Toggle all the debug flags off in kata's configuration.toml to enable full logging.
+	sudo sed -i -e 's/^# *\(enable_debug\).*=.*$/\1 = false/g' "$RUNTIME_CONFIG_PATH"
+}
+
+# Configure containerd for confidential containers. Among other things, it ensures
+# the CRI handler is configured to deal with confidential container.
+#
+# Parameters:
+#	$1 - (Optional) file path to where save the current containerd's config.toml
+#
+configure_cc_containerd() {
+	local saved_containerd_conf_file="${1:-}"
+	local containerd_conf_file="/etc/containerd/config.toml"
+
+	# Even if we are not saving the original file it is a good idea to
+	# restart containerd because it might be in an inconsistent state here.
+	sudo systemctl stop containerd
+	sleep 5
```

Comment:
Why is a sleep required after containerd is stopped? Either way, this is going to be fragile. Can you use `WaitForProcess()` or similar here too?

Reply:
We were using the sleep to ensure the stop process had finished, but using the `WaitForProccess()` one makes more sense so we've switched to use that now instead 🙂